finality-aleph/src/justification.rs

```diff
@@ -29,26 +27,49 @@ pub struct AlephJustification {
 }
 
 impl AlephJustification {
-    pub(crate) fn new<Block: BlockT>(signature: SignatureSet<Signature>) -> Self {
-        Self { signature }
-    }
-
     pub(crate) fn verify<Block: BlockT>(
-        aleph_justification: &AlephJustification,
+        &self,
         block_hash: Block::Hash,
         multi_verifier: &AuthorityVerifier,
     ) -> bool {
-        if !multi_verifier.is_complete(&block_hash.encode()[..], &aleph_justification.signature) {
-            debug!(target: "afa", "Bad justification for block hash #{:?} {:?}", block_hash, aleph_justification);
+        if !multi_verifier.is_complete(&block_hash.encode()[..], &self.signature) {
+            debug!(target: "afa", "Bad justification for block hash #{:?} {:?}", block_hash, self);
             return false;
         }
         true
     }
 }
 
-pub(crate) struct ChainCadence {
-    pub session_period: SessionPeriod,
-    pub justifications_cadence: Duration,
+/// Decides whether enough time has elapsed for making another request for justification.
+pub(crate) trait JustificationRequestDelay {
+    fn can_request_now(&self, last_request_time: Instant, last_finalization_time: Instant) -> bool;
+}
+
+impl<F: Fn(Instant, Instant) -> bool> JustificationRequestDelay for F {
+    fn can_request_now(&self, last_request_time: Instant, last_finalization_time: Instant) -> bool {
+        self(last_request_time, last_finalization_time)
+    }
+}
+
+pub(crate) struct SessionInfo<B: BlockT> {
+    pub(crate) current_session: SessionId,
+    pub(crate) last_block_height: NumberFor<B>,
+    pub(crate) verifier: Option<AuthorityVerifier>,
+}
+
+/// Returns `SessionInfo` for the session regarding block with no. `number`.
+pub(crate) trait SessionInfoProvider<B: BlockT> {
+    fn for_block_num(&self, number: NumberFor<B>) -> SessionInfo<B>;
+}
+
+impl<F, B> SessionInfoProvider<B> for F
```

Comment:
This setup for mocking is very smart and tricky, but I feel like it might be confusing. Right now I don't have any ideas how to mock those functions without such tricks. @DamianStraszak do you have any ideas?

Reply:
there is only `SessionInfoProvider` left; I can remove the blanket implementation for closures and just create a normal struct in `party.rs` if you think that it would increase legibility. I just miss something like functional interfaces from Java :P

Reply:
I'm fine with either version, but there is indeed some trickiness in the current solution that beginner rusteceans might have hard time enjoying :)